src/lib/format/l20n/parser.js

```diff
@@ -224,153 +149,332 @@ export default {
         this.getWS();
       }
+      if (this._source[this._index] === '}') {
+        ++this._index;
+        break;
+      }
+      if (!comma) {
+        throw this.error('Expected "}"');
+      }
+    }
+
+    if (defKey) {
+      items.__default = defKey;
     }
 
-    return ast;
+    return items;
   },
 
-  getIndex: function() {
+  getHashItem: function() {
+    let defItem = false;
+    if (this._source[this._index] === '*') {
+      ++this._index;
+      defItem = true;
+    }
+
+    const key = this.getIdentifier();
     this.getWS();
-    this._patterns.index.lastIndex = this._index;
-    const match = this._patterns.index.exec(this._source);
-    this._index = this._patterns.index.lastIndex;
+    if (this._source[this._index] !== ':') {
+      throw this.error('Expected ":"');
+    }
+    ++this._index;
     this.getWS();
-    this._index++;
 
-    return [{t: 'idOrVar', v: 'plural'}, match[1]];
+    return [key, this.getValue(), defItem];
   },
 
-  parseString: function(str) {
-    const chunks = str.split(this._patterns.placeables);
-    const complexStr = [];
-
-    const len = chunks.length;
-    const placeablesCount = (len - 1) / 2;
+  getComment: function() {
+    this._index += 2;
+    const start = this._index;
+    const end = this._source.indexOf('*/', start);
 
-    if (placeablesCount >= MAX_PLACEABLES) {
-      throw new L10nError('Too many placeables (' + placeablesCount +
-                          ', max allowed is ' + MAX_PLACEABLES + ')');
+    if (end === -1) {
+      throw this.error('Comment without a closing tag');
     }
 
-    for (let i = 0; i < chunks.length; i++) {
-      if (chunks[i].length === 0) {
-        continue;
-      }
-      if (i % 2 === 1) {
-        complexStr.push({t: 'idOrVar', v: chunks[i]});
+    this._index = end + 2;
+  },
+
+  getExpression: function () {
+    let exp = this.getPrimaryExpression();
+
+    while (true) {
+      let ch = this._source[this._index];
+      if (ch === '.' || ch === '[') {
+        ++this._index;
+        exp = this.getPropertyExpression(exp, ch === '[');
+      } else if (ch === '(') {
+        ++this._index;
+        exp = this.getCallExpression(exp);
       } else {
-        complexStr.push(chunks[i]);
+        break;
       }
     }
-    return complexStr;
+
+    return exp;
   },
 
-  error: function(message, pos) {
-    if (pos === undefined) {
-      pos = this._index;
+  getPropertyExpression: function(idref, computed) {
+    let exp;
+
+    if (computed) {
+      this.getWS();
+      exp = this.getExpression();
+      this.getWS();
+      if (this._source[this._index] !== ']') {
+        throw this.error('Expected "]"');
+      }
+      ++this._index;
+    } else {
+      exp = this.getIdentifier();
     }
+
+    return {
+      type: 'prop',
+      expr: idref,
+      prop: exp,
+      cmpt: computed
+    };
+  },
+
+  getCallExpression: function(callee) {
+    this.getWS();
+
+    return {
+      type: 'call',
+      call: callee,
```

Comment:
I know I was the one who suggested it, but I think `call` is confusing because of `Function.prototype.call`.  Maybe `expr`, similar to `PropertyExpression`?